nixos/lib/testing: add separate configuration option for composability

ethancedwards8 opened the pull request:
Ready for feedback!


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

fricklerhandwerk commented on nixos/tests/stalwart/stalwart-mail.nix:
```diff
@@ -10,13 +10,10 @@ in
 {
   name = "stalwart-mail";
 
+  nodes.main.configuration = import ./stalwart-mail-config.nix;
```
```suggestion
  nodes.main.configuration = ./stalwart-mail-config.nix;
```

fricklerhandwerk commented on nixos/lib/testing/test-configuration.nix:
```diff
@@ -0,0 +1,23 @@
+    description = "Additional configuration to be passed to the test";
+    default = { };
+    type = types.attrsOf (
+      types.submodule (
```
```suggestion
      types.deferredModule {
```

```suggestion
      types.deferredModuleWith { staticModules = [ ]; }
```

ethancedwards8 commented:
Note that I chose `types.raw` because `deferredModule` hides the filepath behind an imports that does not give us what we need.

ethancedwards8 committed:
nixos/lib/testing: add separate configuration option for composability

Need in NGIpkgs to access configurations.

Can be accessed by:
    nix-repl> nixosTests.<test>.config.configurations.<node>

Co-authored-by: Luj <julien@malka.sh>
Signed-off-by: Ethan Carter Edwards <ethan@ethancedwards.com>

ethancedwards8 committed:
nixos/tests/stalwart: migrate to new configurations option

Can be accessed by:
    nix-repl> nixosTests.stalwart-mail.config.configurations.main

Signed-off-by: Ethan Carter Edwards <ethan@ethancedwards.com>

ethancedwards8 committed:
nixos/tests/alps: migrate to new configurations option

This allows us to access alps configurations through:
nixosTests.alps.config.configurations.server

And it can be imported by downstream users of Nixpkgs (like NGIpkgs).

Signed-off-by: Ethan Carter Edwards <ethan@ethancedwards.com>

ethancedwards8 commented:
Several important maintainers have expressed concerns about this approach. Marking as a draft for now until we find a solution that works for everyone.
